Allow to selectively copy plugins

jeremyVignelles opened the pull request:
With this feature, you can select which plugins gets copied/not copied (see README). 🎉 

The use of batching requires that we use a Task for that, that executes before build. As a side effect, Visual Studio doesn't show the libvlc folder in the solution folder anymore! 🎉 again!

mfkl commented on README.md:
```diff
@@ -79,9 +79,10 @@ To-do:
 
 # How do I configure what gets copied to my output directory?
```
Probably should warn before this line (80) that the following is about Windows (and WindowsRT) only.

mfkl replied:
Or create a new page for platform specific info? What do you think?

jeremyVignelles replied:
As discussed in private : It's not just about windows, it will be about linux too, if it's done one day. I'm ok to split the files, I'm doing it now, but by topic rather than platform.

mfkl commented on README.md:
```diff
@@ -126,3 +127,66 @@ Example : put libvlc to `native/x86`/`native/x64`
+`VlcExcludeWindowsPlugins` item group.
+That exclusion will apply to all windows builds of libvlc (x86 and x64).
+
+There is not such inclusion/exclusion mechanism for Android because libvlc is built as one monolithic library on this platform.
```
Not just Android, you can basically say "the other supported platforms"

mfkl commented on README.md:
```diff
@@ -126,3 +127,66 @@ Example : put libvlc to `native/x86`/`native/x64`
+
+A few things to note:
+
+- It's in an `ItemGroup`, not in a `PropertyGroup` as before
```
What is "it"? What is "before"? I think this could be clearer.

jeremyVignelles replied:
Removed that statement, wasn't really useful as long as you pay attention to what is written (or what you copy/paste)

mfkl commented on README.md:
```diff
@@ -126,3 +127,66 @@ Example : put libvlc to `native/x86`/`native/x64`
+Sometimes, you want to build a minimal package for one of these reasons:
+- Reduce your package size
+- Remove attacking surface of your software by limiting its features
+- Speed up build times : Fastest copies are those that do not occur.
```
Please remove the space before the colon (:).

mfkl commented on README.md:
```diff
@@ -126,3 +127,66 @@ Example : put libvlc to `native/x86`/`native/x64`
+- Reduce your package size
+- Remove attacking surface of your software by limiting its features
+- Speed up build times : Fastest copies are those that do not occur.
+- Just because you don't need that plugin, it's cleaner not to copy it.
```
Please avoid contractions when possible. "You do not".

mfkl commented on README.md:
```diff
@@ -126,3 +127,66 @@ Example : put libvlc to `native/x86`/`native/x64`
+A few things to note:
+
+- It's in an `ItemGroup`, not in a `PropertyGroup` as before
+- Even if it's an exclude, we use `Include` to choose which plugins to exclude
```
I read that sentence 3 times before I understood it ^^

Basically what you want to say is that no matter what we do, we use the `Include` keyword from MSBuild. Conveying that might be confusing, so why not use the `Exclude` MSBuild keyword that'd match our intent?
https://msdn.microsoft.com/en-us/library/ms171453.aspx#BKMK_ExcludeAttribute
Why not use it?

jeremyVignelles replied:
We're building a string list, in which you "include" items. It doesn't have any special meaning for MSBuild (as long as you are careful to escape the `*`). Afterwards, the `%2A` are converted into `*`, which is interpreted as msbuild.

In other words:

```
<Something Include="a;b;c" Exclude="b" />
```

will result in the list `a;c`

but

```
  <SomethingElse Include="a/%2A;b" Exclude="a/foo" />
```

will result in the list `a/*;b`, and `a/foo` won't be excluded (since that string is not in the list)

mfkl commented on README.md:
```diff
@@ -126,3 +127,66 @@ Example : put libvlc to `native/x86`/`native/x64`
+
+  <!-- You can exclude with wildcards -->
+  <VlcExcludeWindowsPlugins Include="%2A/%2Adummy%2A" />
+</ItemGroup>
```
good examples 👍

mfkl commented on README.md:
````diff
@@ -126,3 +127,66 @@ Example : put libvlc to `native/x86`/`native/x64`
+</ItemGroup>
+```
+
+## Include selectively items
````
-> Selectively Include plugins
or "cherry pick the plugins you need"

mfkl commented on README.md:
````diff
@@ -126,3 +127,66 @@ Example : put libvlc to `native/x86`/`native/x64`
+```
+
+## Include selectively items
+Another solution to achieve the goals mentioned above is to copy only the plugins that you want in your build.
````
You may also use the following inclusive strategy to reach the goals defined above.

mfkl commented:
>  As a side effect, Visual Studio doesn't show the libvlc folder in the solution folder anymore! 

Now I'm wondering. Is that a good thing or not?

jeremyVignelles commented:
>>    As a side effect, Visual Studio doesn't show the libvlc folder in the solution folder anymore!
>
> Now I'm wondering. Is that a good thing or not?

I think it is:
- Avoids confusion in the project folder ("What's that?", "Oops, I deleted that by accident"...)
- I suspect faster load times for the project
- What happens when you look for a string in the whole solution? Will those files be scanned?

We don't really have the choice anyway: batching is a powerful feature that would be hard to replace in a pure-msbuild way.

Anyway, I updated my branch based on your comments, I'm waiting for your review 😄

mfkl commented:
So what's the status for this? Is it ready on your side?

You include everything by default right? hrtf etc

jeremyVignelles commented:
It's not ready : I will split win32/64 selections and explain how to do accordingly. Including other files in the nuget package is out of scope of this PR, and you are welcome to do that ( and I will rebase my work on top of that, and include the other things by default when needed)

jeremyVignelles commented:
Finished adding files to the package (lua, locales and hrtfs). 
Don't know what languages is, so I didn't include it.

for me, it's ready for a merge :)

mfkl commented:
Built a nuget with your branch and added the following lines my sample project's csproj:
```
  <VlcWindowsX64ExcludeFiles Include="plugins/gui/libqt_plugin.dll" />
  <VlcWindowsX64ExcludeFiles Include="plugins/lua" />
  <VlcWindowsX86ExcludeFiles Include="@(VlcWindowsX64ExcludeFiles)" />
```

Everything still gets copied. What am I missing here?

jeremyVignelles commented:
Did you do that in an ItemGroup? did you build a new version or did you forget to remove the cache?

mfkl commented:
> Did you do that in an ItemGroup?

Yes.

> did you build a new version or did you forget to remove the cache?

New alpha version.

I guess you don't have this issue..

jeremyVignelles commented:
I have to check again, but as far as I remember, I didn't have that issue

jeremyVignelles commented:
For lua, you need to select files, not the directory IIRC. Try appending `/%2A%2A`

jeremyVignelles commented:
Folder separator must always be the same. Since I have no control over what is generated with globbing (`**`gets expanded to `a\b\c`), I decided to use `\` everywhere, until this issue is resolved : https://github.com/Microsoft/msbuild/issues/1024

You cannot select a folder with include or exclude, use `\%2A` at the end to select all its files or `\%2A%2A` to select recursively all the files at any level under.

Sorry for the lack of testing, it should be better now, but you are welcome to check twice ;)

mfkl commented:
Thanks for your edits. It works here. 2 remarks:
- I think it would be relevant to mention in the doc that you may use the inclusive or the exclusive strategy, but not *both* at the same time.
- What do you think about printing a message in the console output displaying the count of included/excluded plugins at build time?

jeremyVignelles commented:
You can perfectly use both strategy. in fact, that's what you're doing when you're just using  Exclude : there is a default include.

Printing a message could be done, but requires me to add temp items steps in the .targets. Is that really something useful? If I remember correctly, if you increase msbuild verbosity, you can have the full list of what gets copied (including your own project's DLL)

mfkl commented:
> Is that really something useful?

Yes, I definitely think it would be. 

It's informative for people who just added the nuget along with C# bindings, are not familiar with libvlc and did not read the docs. It would print something like `VLC plugins included: 412 / VLC plugins excluded: 2` in the VS console, which *suggests* you can actually configure that. Maybe even print a link to your cherry-pick readme page.

Passing `/v:diag` to `msbuild` does the trick, although it makes the output unreadable if you include most plugins as it prints the names of the plugin dlls instead of just the count. Maybe we could do something with https://docs.microsoft.com/en-us/visualstudio/msbuild/build-loggers later on. That can wait though.

mfkl commented:
https://www.nuget.org/packages/VideoLAN.LibVLC.Windows/3.0.0-alpha3
